win32ss/user/winsrv/consrv/condrv/text.c

```diff
@@ -932,138 +933,273 @@ ConDrvReadConsoleOutputString(IN PCONSOLE Console,
     }
 }
 
-NTSTATUS NTAPI
-ConDrvWriteConsoleOutputString(IN PCONSOLE Console,
-                               IN PTEXTMODE_SCREEN_BUFFER Buffer,
-                               IN CODE_TYPE CodeType,
-                               IN PVOID StringBuffer,
-                               IN ULONG NumCodesToWrite,
-                               IN PCOORD WriteCoord,
-                               // OUT PCOORD EndCoord,
-                               OUT PULONG NumCodesWritten OPTIONAL)
+static NTSTATUS
+IntWriteConsoleOutputStringAscii(
+    IN PCONSOLE Console,
+    IN PTEXTMODE_SCREEN_BUFFER Buffer,
+    IN PVOID StringBuffer,
+    IN ULONG NumCodesToWrite,
+    IN PCOORD WriteCoord,
+    OUT PULONG NumCodesWritten OPTIONAL)
 {
     NTSTATUS Status = STATUS_SUCCESS;
-    PVOID WriteBuffer = NULL;
+    LPBYTE WriteBuffer;
     PWCHAR tmpString = NULL;
-    ULONG X, Y, Length; // , Written = 0;
-    ULONG CodeSize;
+    ULONG i, X, Y, Length;
     PCHAR_INFO Ptr;
+    BOOL bCJK = Console->IsCJK;
 
-    if (Console == NULL || Buffer == NULL || WriteCoord == NULL /* || EndCoord == NULL */)
+    if (!StringBuffer)
+        goto Cleanup;
+
+    /* Convert the ASCII string into Unicode before writing it to the console */
+    Length = MultiByteToWideChar(Console->OutputCodePage, 0,
+                                 (PCHAR)StringBuffer,
+                                 NumCodesToWrite,
+                                 NULL, 0);
+    tmpString = RtlAllocateHeap(RtlGetProcessHeap(), 0, Length * sizeof(WCHAR));
+    if (tmpString)
     {
-        return STATUS_INVALID_PARAMETER;
+        WriteBuffer = (LPBYTE)tmpString;
+        MultiByteToWideChar(Console->OutputCodePage, 0,
+                            (PCHAR)StringBuffer,
+                            NumCodesToWrite,
+                            (PWCHAR)WriteBuffer, Length);
+    }
+    else
+    {
+        Status = STATUS_NO_MEMORY;
+        goto Cleanup;
     }
 
-    /* Validity checks */
-    ASSERT(Console == Buffer->Header.Console);
-    ASSERT((StringBuffer != NULL) || (StringBuffer == NULL && NumCodesToWrite == 0));
+    X = WriteCoord->X;
```

Comment:
Is there a difference between all this code below and `IntWriteConsoleOutputStringUnicode`? If no, let's just call the other function instead of duplicating the code.

Reply:
OK.